### PR TITLE
FIX PiSSA & OLoRA with rank/alpha pattern, rslora

### PR DESCRIPTION
**Pull request by BenjaminBossan**
See https://github.com/huggingface/peft/issues/1929#issuecomment-2230780802

At the moment, when using PiSSA or OLoRA with weight conversion to restore the original base weights, there is an error when either of `rank_pattern`, `alpha_pattern`, or `rslora` is being used. This PR fixes this.

The issue is that we need to double the rank of the LoRA adapter. Right now, this is done by simply doubling r and alpha. But if `rank_pattern` and `alpha_pattern` are being used, those need to be doubled too.

Furthermore, when using `rslora`, the scaling is again different, namely `alpha / sqrt(r)`. This also needs to be adjusted.

Unfortunately, when using `rslora` together with `rank_pattern` and `alpha_pattern`, this gets way more complicated. Since we don't store the scaling in the `state_dict`, we would have to resolve all the patterns here to determine the correct scaling, i.e. reimplement the whole matching and init logic. This is a lot of work for a very edgy edge case.

Therefore, I opted to raise an error instead. This is not super nice, as the error is only raised when trying to save the model, i.e. a lot of time may already have been spent to train the model. But we cannot know this earlier, so not much can be done.

Overall, this fix is ugly because it further couples unrelated code. For instance, if we add new init methods that affect the scaling, we need to remember to change the saving logic accordingly. If anyone has a better idea, LMK.

## Update

I noticed that the test `test_lora_pissa_conversion_same_output_after_loading_with_quantization` was failing locally. This is because the argument `convert_mutated_to_lora` was renamed to `path_initial_model_for_weight_conversion` in #1828. This oversight has been addressed. However, this also means that the tests did not run in CI (nightly CI with GPU). The reason was that the PiSSA and OLoRA tests were missing the right marker, so I added them as well.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Tagging @fxmeng and @tokenizer-decode as this affects the code they contributed.

**Comment by tokenizer-decode** [comment]
Oh, so we raise the error after all the training is done? Can you tell me why we are checking `peft_config.use_rslora and (peft_config.rank_pattern or peft_config.alpha_pattern):` after all the training is and before trying the save the model? Why are we not doing this when initializing PiSSA or OLoRA? Aren't `peft_config.rank_pattern ` and `peft_config.alpha_pattern` options determined before doing anything? Or maybe I did not understand you correctly.

**Comment by tokenizer-decode** [comment]
Hmm. Maybe at the initialization if  `peft_config.use_rslora and (peft_config.rank_pattern or peft_config.alpha_pattern) ` is `True` we can warn the user saying you cannot use `path_initial_model_for_weight_conversion`? There is a chance he may stop there. I don't know. This is also not a good fix but at least gives user a chance. I would be pissed after spending bunch of many to an EC2 instance then getting this error.

**Comment by BenjaminBossan** [comment]
@tokenizer-decode Good suggestion about adding a warning. I did that in the last commit, though unfortunately many users don't read warnings. Still, there is not much else that can be done here.

**Comment by tokenizer-decode** [comment]
Yeah you are right not much else to do here. But I agree that this is a very edgy edge case so we should be ok.

**Comment by BenjaminBossan** [comment]
@sayakpaul LMK if you don't want to review this, I'll look for someone else.

**Comment by BenjaminBossan** [comment]
@tokenizer-decode While working on this, I also discovered a small error in a test caused by your PR, which I have fixed (see **update** section of PR description).

**Comment by tokenizer-decode** [comment]
Oh I totally missed that. Thanks

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -258,6 +258,13 @@ def save_pretrained(
             path_initial_model_for_weight_conversion = convert_pissa_to_lora
 
         def save_mutated_as_lora(peft_config, path_initial_model_for_weight_conversion, output_state_dict, kwargs):
+            if peft_config.use_rslora and (peft_config.rank_pattern or peft_config.alpha_pattern):
```
Comment (sayakpaul) [inline src/peft/peft_model.py:261]: Can't we use this condition as soon as this permutation of config is requested and warn/error?

File: src/peft/tuners/lora/config.py [hunk 2]
```diff
@@ -346,6 +347,26 @@ def __post_init__(self):
...
+                or (self.init_lora_weights == "olora")
+            )
+        ):
+            msg = (
+                "Using Rank-Stabilized LoRA with rank_pattern/alpha_pattern and post-training conversion of modified "
+                "base weights (PiSSA, OLoRA) means that you won't be able to pass "
+                "`path_initial_model_for_weight_conversion` to `save_pretrained` to restore the initial values of the "
+                "base weights; if you intend to do this, please ensure not to use rslora or rank_pattern/alpha_pattern."
+            )
+            warnings.warn(msg)
```
Comment (sayakpaul) [inline src/peft/tuners/lora/config.py:368]: Maybe we can document this limitation in the API spec itself, as well? 

I would also remember to include this in the release notes so that users are firmly aware of this.

File: tests/test_initialization.py [hunk 3]
```diff
@@ -338,6 +338,194 @@ def test_lora_pissa_conversion_same_output_after_loading(self, data, tmp_path):
...
+        peft_model.peft_config["default"].init_lora_weights = True
+        peft_model.save_pretrained(tmp_path / "init-model")
+        peft_model.peft_config["default"].init_lora_weights = "pissa"
+
+        # modify the weights, or else the adapter performs an identity transformation
+        peft_model.base_model.linear.lora_B["default"].weight.data *= 2.0
+        output_pissa = peft_model(data)[0]
+
+        # sanity check
+        tol = 1e-06
```
Comment (sayakpaul) [inline tests/test_initialization.py:359]: (nit) a stricter check for negations is to relax the tolerance a bit and see if that still holds IMO.

File: tests/test_initialization.py [hunk 4]
```diff
@@ -338,6 +338,194 @@ def test_lora_pissa_conversion_same_output_after_loading(self, data, tmp_path):
...
+
+        # save the model normally
+        peft_model.save_pretrained(tmp_path / "pissa-model")
+        model_loaded = PeftModel.from_pretrained(deepcopy(model), tmp_path / "pissa-model")
+        output_loaded = model_loaded(data)[0]
+
+        assert torch.allclose(output_pissa, output_loaded, atol=tol, rtol=tol)
+        # sanity check: ranks should still be 8 as initially
+        assert model_loaded.peft_config["default"].r == 8
+        assert model_loaded.base_model.model.linear.lora_A["default"].weight.shape[0] == 32
```
Comment (sayakpaul) [inline tests/test_initialization.py:370]: What is the difference between `r` being 8 and `rank_pattern` being 32 for `linear` here? 

`target_modules=["linear"], r=8, rank_pattern={"linear": 32}`

File: tests/test_initialization.py [hunk 5]
```diff
@@ -338,6 +338,194 @@ def test_lora_pissa_conversion_same_output_after_loading(self, data, tmp_path):
...
+        model = self.get_model()
+        output_base = model(data)[0]
+
+        # use rank_pattern here
+        config = LoraConfig(init_lora_weights="pissa", target_modules=["linear"], r=8, rank_pattern={"linear": 32})
+        peft_model = get_peft_model(deepcopy(model), config)
+        # save the initial model
+        peft_model.peft_config["default"].init_lora_weights = True
+        peft_model.save_pretrained(tmp_path / "init-model")
+        peft_model.peft_config["default"].init_lora_weights = "pissa"
```
Comment (sayakpaul) [inline tests/test_initialization.py:352]: Help me unpack this bit of code.

We are using `pissa` for `init_lora_weights` already in the `LoraConfig`. So, I don't understand what would differ between `peft_model.peft_config["default"].init_lora_weights = True` and `peft_model.peft_config["default"].init_lora_weights = "pissa"`.

File: tests/test_initialization.py [hunk 6]
```diff
@@ -338,6 +338,194 @@ def test_lora_pissa_conversion_same_output_after_loading(self, data, tmp_path):
...
+        assert torch.allclose(output_pissa, output_converted, atol=tol, rtol=tol)
+        # rank should be double of what it was initially
+        assert model_converted.peft_config["default"].r == 16
+        assert model_converted.base_model.model.linear.lora_A["default"].weight.shape[0] == 64
+        # base model weights should be the same as the initial model
+        assert torch.allclose(
+            model.linear.weight, model_converted.base_model.model.linear.base_layer.weight, atol=tol, rtol=tol
+        )
+
+    def test_lora_pissa_conversion_same_output_after_loading_with_alpha_pattern(self, data, tmp_path):
```
Comment (sayakpaul) [inline tests/test_initialization.py:392]: Same comments apply for this too.

**Review by sayakpaul — COMMENTED** [review]
Thanks! Seems like a very deep edge case. I have left some comments to better understand some stuff. Once replied to, I will review one more time.

File: tests/test_initialization.py (hunk 5 above)
Comment (tokenizer-decode) [reply tests/test_initialization.py:352]: That was due to a bug in original PiSSA implementation. I fixed it when I was implementing olora but I didn’t change tests. What would differ is `init_lora_weights `should be `True` in the config file. See [here](https://github.com/huggingface/peft/pull/1828#issuecomment-2160483919)

File: src/peft/peft_model.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/peft_model.py:261]: Yes, this is done below and we warn the user: https://github.com/huggingface/peft/pull/1930/files#diff-13380145401d203d5935c5189dd09879f990b81aa63e8e3aaff8ce9110333f0eR354

We don't raise an error since we cannot know at that point in time if they'll use the `path_initial_model_for_weight_conversion`. If they don't, there is no problem.

File: src/peft/tuners/lora/config.py (hunk 2 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/config.py:368]: I'll add a mention to the docstring of `save_pretrained`, as well as to the README files of the two methods.

I added an entry to the PEFT release notes to mention this change there as well. As OLoRA is as yet unreleased and PiSSA was released in the last version, and also given that this is an edge case, I think very few users, if any, will be affected.

File: tests/test_initialization.py (hunk 3 above)
Comment (BenjaminBossan) [reply tests/test_initialization.py:359]: True. At the moment, we use the same tolerance for `assert not torch.allclose(...)` and `assert torch.allclose(...)` throughout the whole PEFT code base. Therefore, I wouldn't change this now only for these specific tests. I think this could be a nice use case for a code-LLM to go through all tests and rewrite them to use higher tolerance for negative checks :)

File: tests/test_initialization.py (hunk 4 above)
Comment (BenjaminBossan) [reply tests/test_initialization.py:370]: Sorry, I missed this comment earlier, so addressing it now.

For this simple test, we only have a small model with a single linear layer. Therefore, if we set `rank_pattern={"linear": 32}`, we basically ignore `r=8` because `rank_pattern` takes precedence. If we had a real model, we could, however, set all ranks to 8 this way and pick individual layers with a different rank by using `rank_pattern`.

Does that answer the question?

File: src/peft/peft_model.py (hunk 1 above)
Comment (sayakpaul) [reply src/peft/peft_model.py:263]: Okay that should cut it.

File: tests/test_initialization.py (hunk 4 above)
Comment (sayakpaul) [reply tests/test_initialization.py:370]: Yeah it does. Maybe just add a comment explaining this in the test? Because many developers refer to tests to understand different usage patterns and they may get confused here as well.

**Review by sayakpaul — APPROVED** [review]
Very minor changes requested. I think this is the best we can have right now.

**Comment by BenjaminBossan** [comment]
@sayakpaul I added clarifying comments to the tests in my latest commit.
